### ISSUE TITLE
argon2-cffi fail while installing jupyter 

### ISSUE THREAD

**Issue by olav2021** [termux/termux-packages#7925]
### Problem description

Hello,

trying to get jpyter up and running on an arm81 , argon2-cffi 21.1.0 installation throws error "command llvm-arm failed : No such file or directory"

~ $ pip3 install jupyter --ignore-installed argon2-cffi
Collecting jupyter
  Using cached jupyter-1.0.0-py2.py3-none-any.whl (2.7 kB)
Collecting argon2-cffi
  Using cached argon2-cffi-21.1.0.tar.gz (1.8 MB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Collecting ipykernel
  Using cached ipykernel-6.5.0-py3-none-any.whl (125 kB)
Collecting nbconvert
  Using cached nbconvert-6.2.0-py3-none-any.whl (553 kB)
Collecting jupyter-console
  Using cached jupyter_console-6.4.0-py3-none-any.whl (22 kB)
Collecting notebook
  Using cached notebook-6.4.5-py3-none-any.whl (9.9 MB)
Collecting ipywidgets
  Using cached ipywidgets-7.6.5-py2.py3-none-any.whl (121 kB)
Collecting qtconsole
  Using cached qtconsole-5.1.1-py3-none-any.whl (119 kB)
Collecting cffi>=1.0.0
  Using cached cffi-1.15.0.tar.gz (484 kB)
  Preparing metadata (setup.py) ... done
Collecting pycparser
  Using cached pycparser-2.20-py2.py3-none-any.whl (112 kB)
Collecting tornado<7.0,>=4.2
  Using cached tornado-6.1.tar.gz (497 kB)
  Preparing metadata (setup.py) ... done
Collecting jupyter-client<8.0
  Using cached jupyter_client-7.0.6-py3-none-any.whl (125 kB)
Collecting traitlets<6.0,>=5.1.0
  Using cached traitlets-5.1.1-py3-none-any.whl (102 kB)
Collecting ipython<8.0,>=7.23.1
  Using cached ipython-7.29.0-py3-none-any.whl (790 kB)
Collecting matplotlib-inline<0.2.0,>=0.1.0
  Using cached matplotlib_inline-0.1.3-py3-none-any.whl (8.2 kB)
Collecting debugpy<2.0,>=1.0.0
  Using cached debugpy-1.5.1-py2.py3-none-any.whl (4.1 MB)
Collecting widgetsnbextension~=3.5.0
  Using cached widgetsnbextension-3.5.2-py2.py3-none-any.whl (1.6 MB)
Collecting jupyterlab-widgets>=1.0.0
  Using cached jupyterlab_widgets-1.0.2-py3-none-any.whl (243 kB)
Collecting nbformat>=4.2.0
  Using cached nbformat-5.1.3-py3-none-any.whl (178 kB)
Collecting ipython-genutils~=0.2.0
  Using cached ipython_genutils-0.2.0-py2.py3-none-any.whl (26 kB)
Collecting pygments
  Using cached Pygments-2.10.0-py3-none-any.whl (1.0 MB)
Collecting prompt-toolkit!=3.0.0,!=3.0.1,<3.1.0,>=2.0.0
  Using cached prompt_toolkit-3.0.22-py3-none-any.whl (374 kB)
Collecting nbclient<0.6.0,>=0.5.0
  Using cached nbclient-0.5.4-py3-none-any.whl (66 kB)
Collecting pandocfilters>=1.4.1
  Using cached pandocfilters-1.5.0-py2.py3-none-any.whl (8.7 kB)
Collecting jupyterlab-pygments
  Using cached jupyterlab_pygments-0.1.2-py2.py3-none-any.whl (4.6 kB)
Collecting mistune<2,>=0.8.1
  Using cached mistune-0.8.4-py2.py3-none-any.whl (16 kB)
Collecting defusedxml
  Using cached defusedxml-0.7.1-py2.py3-none-any.whl (25 kB)
Collecting jinja2>=2.4
  Using cached Jinja2-3.0.2-py3-none-any.whl (133 kB)
Collecting bleach
  Using cached bleach-4.1.0-py2.py3-none-any.whl (157 kB)
Collecting jupyter-core
  Using cached jupyter_core-4.9.1-py3-none-any.whl (86 kB)
Collecting testpath
  Using cached testpath-0.5.0-py3-none-any.whl (84 kB)
Collecting entrypoints>=0.2.2
  Using cached entrypoints-0.3-py2.py3-none-any.whl (11 kB)
Collecting pyzmq>=17
  Using cached pyzmq-22.3.0-cp310-cp310-linux_armv8l.whl
Collecting prometheus-client
  Using cached prometheus_client-0.12.0-py2.py3-none-any.whl (57 kB)
Collecting Send2Trash>=1.5.0
  Using cached Send2Trash-1.8.0-py3-none-any.whl (18 kB)
Collecting terminado>=0.8.3
  Using cached terminado-0.12.1-py3-none-any.whl (15 kB)
Collecting qtpy
  Using cached QtPy-1.11.2-py2.py3-none-any.whl (58 kB)
Collecting decorator
  Using cached decorator-5.1.0-py3-none-any.whl (9.1 kB)
Collecting backcall
  Using cached backcall-0.2.0-py2.py3-none-any.whl (11 kB)
Collecting pickleshare
  Using cached pickleshare-0.7.5-py2.py3-none-any.whl (6.9 kB)
Collecting pexpect>4.3
  Using cached pexpect-4.8.0-py2.py3-none-any.whl (59 kB)
Collecting jedi>=0.16
  Using cached jedi-0.18.0-py2.py3-none-any.whl (1.4 MB)
Collecting setuptools>=18.5
  Using cached setuptools-58.5.2-py3-none-any.whl (946 kB)
Collecting MarkupSafe>=2.0
  Using cached MarkupSafe-2.0.1.tar.gz (18 kB)
  Preparing metadata (setup.py) ... done
Collecting nest-asyncio>=1.5
  Using cached nest_asyncio-1.5.1-py3-none-any.whl (5.0 kB)
Collecting python-dateutil>=2.1
  Using cached python_dateutil-2.8.2-py2.py3-none-any.whl (247 kB)
Collecting jsonschema!=2.5.0,>=2.4
  Using cached jsonschema-4.2.0-py3-none-any.whl (69 kB)
Collecting wcwidth
  Using cached wcwidth-0.2.5-py2.py3-none-any.whl (30 kB)
Collecting ptyprocess
  Using cached ptyprocess-0.7.0-py2.py3-none-any.whl (13 kB)
Collecting packaging
  Using cached packaging-21.2-py3-none-any.whl (40 kB)
Collecting six>=1.9.0
  Using cached six-1.16.0-py2.py3-none-any.whl (11 kB)
Collecting webencodings
  Using cached webencodings-0.5.1-py2.py3-none-any.whl (11 kB)
Collecting parso<0.9.0,>=0.8.0
  Using cached parso-0.8.2-py2.py3-none-any.whl (94 kB)
Collecting pyrsistent!=0.17.0,!=0.17.1,!=0.17.2,>=0.14.0
  Using cached pyrsistent-0.18.0-cp310-cp310-linux_armv8l.whl
Collecting attrs>=17.4.0
  Using cached attrs-21.2.0-py2.py3-none-any.whl (53 kB)
Collecting pyparsing<3,>=2.0.2
  Using cached pyparsing-2.4.7-py2.py3-none-any.whl (67 kB)
Using legacy 'setup.py install' for cffi, since package 'wheel' is not installed.
Using legacy 'setup.py install' for tornado, since package 'wheel' is not installed.
Using legacy 'setup.py install' for MarkupSafe, since package 'wheel' is not installed.
Building wheels for collected packages: argon2-cffi
  Building wheel for argon2-cffi (pyproject.toml) ... error
  ERROR: Command errored out with exit status 1:
   command: /data/data/com.termux/files/usr/bin/python3 /data/data/com.termux/files/usr/lib/python3.10/site-packages/pip/_vendor/pep517/in_process/_in_process.py build_wheel /data/data/com.termux/files/usr/tmp/tmpq0huuqtm
       cwd: /data/data/com.termux/files/usr/tmp/pip-install-nxpytfh_/argon2-cffi_035c8e11597e4541b8fb07a4dd518867
  Complete output (29 lines):
  running bdist_wheel
  running build
  running build_py
  creating build
  creating build/lib.linux-armv8l-3.10
  creating build/lib.linux-armv8l-3.10/argon2
  copying src/argon2/__init__.py -> build/lib.linux-armv8l-3.10/argon2
  copying src/argon2/low_level.py -> build/lib.linux-armv8l-3.10/argon2
  copying src/argon2/_ffi_build.py -> build/lib.linux-armv8l-3.10/argon2
  copying src/argon2/exceptions.py -> build/lib.linux-armv8l-3.10/argon2
  copying src/argon2/_utils.py -> build/lib.linux-armv8l-3.10/argon2
  copying src/argon2/_legacy.py -> build/lib.linux-armv8l-3.10/argon2
  copying src/argon2/_password_hasher.py -> build/lib.linux-armv8l-3.10/argon2
  copying src/argon2/__main__.py -> build/lib.linux-armv8l-3.10/argon2
  running build_clib
  building 'argon2' library
  creating build/temp.linux-armv8l-3.10
  creating build/temp.linux-armv8l-3.10/extras
  creating build/temp.linux-armv8l-3.10/extras/libargon2
  creating build/temp.linux-armv8l-3.10/extras/libargon2/src
  creating build/temp.linux-armv8l-3.10/extras/libargon2/src/blake2
  arm-linux-androideabi-clang -mfloat-abi=softfp -mfpu=vfpv3-d16 -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -march=armv7-a -mfpu=neon -mfloat-abi=softfp -mthumb -fstack-protector-strong -O3 -march=armv7-a -mfpu=neon -mfloat-abi=softfp -mthumb -fstack-protector-strong -O3 -fPIC -Iextras/libargon2/src/../include -Iextras/libargon2/src/blake2 -c extras/libargon2/src/argon2.c -o build/temp.linux-armv8l-3.10/extras/libargon2/src/argon2.o
  arm-linux-androideabi-clang -mfloat-abi=softfp -mfpu=vfpv3-d16 -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -march=armv7-a -mfpu=neon -mfloat-abi=softfp -mthumb -fstack-protector-strong -O3 -march=armv7-a -mfpu=neon -mfloat-abi=softfp -mthumb -fstack-protector-strong -O3 -fPIC -Iextras/libargon2/src/../include -Iextras/libargon2/src/blake2 -c extras/libargon2/src/blake2/blake2b.c -o build/temp.linux-armv8l-3.10/extras/libargon2/src/blake2/blake2b.o
  arm-linux-androideabi-clang -mfloat-abi=softfp -mfpu=vfpv3-d16 -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -march=armv7-a -mfpu=neon -mfloat-abi=softfp -mthumb -fstack-protector-strong -O3 -march=armv7-a -mfpu=neon -mfloat-abi=softfp -mthumb -fstack-protector-strong -O3 -fPIC -Iextras/libargon2/src/../include -Iextras/libargon2/src/blake2 -c extras/libargon2/src/core.c -o build/temp.linux-armv8l-3.10/extras/libargon2/src/core.o
  arm-linux-androideabi-clang -mfloat-abi=softfp -mfpu=vfpv3-d16 -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -march=armv7-a -mfpu=neon -mfloat-abi=softfp -mthumb -fstack-protector-strong -O3 -march=armv7-a -mfpu=neon -mfloat-abi=softfp -mthumb -fstack-protector-strong -O3 -fPIC -Iextras/libargon2/src/../include -Iextras/libargon2/src/blake2 -c extras/libargon2/src/encoding.c -o build/temp.linux-armv8l-3.10/extras/libargon2/src/encoding.o
  arm-linux-androideabi-clang -mfloat-abi=softfp -mfpu=vfpv3-d16 -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -march=armv7-a -mfpu=neon -mfloat-abi=softfp -mthumb -fstack-protector-strong -O3 -march=armv7-a -mfpu=neon -mfloat-abi=softfp -mthumb -fstack-protector-strong -O3 -fPIC -Iextras/libargon2/src/../include -Iextras/libargon2/src/blake2 -c extras/libargon2/src/ref.c -o build/temp.linux-armv8l-3.10/extras/libargon2/src/ref.o
  arm-linux-androideabi-clang -mfloat-abi=softfp -mfpu=vfpv3-d16 -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -march=armv7-a -mfpu=neon -mfloat-abi=softfp -mthumb -fstack-protector-strong -O3 -march=armv7-a -mfpu=neon -mfloat-abi=softfp -mthumb -fstack-protector-strong -O3 -fPIC -Iextras/libargon2/src/../include -Iextras/libargon2/src/blake2 -c extras/libargon2/src/thread.c -o build/temp.linux-armv8l-3.10/extras/libargon2/src/thread.o
  llvm-ar rcs build/temp.linux-armv8l-3.10/libargon2.a build/temp.linux-armv8l-3.10/extras/libargon2/src/argon2.o build/temp.linux-armv8l-3.10/extras/libargon2/src/blake2/blake2b.o build/temp.linux-armv8l-3.10/extras/libargon2/src/core.o build/temp.linux-armv8l-3.10/extras/libargon2/src/encoding.o build/temp.linux-armv8l-3.10/extras/libargon2/src/ref.o build/temp.linux-armv8l-3.10/extras/libargon2/src/thread.o
  error: command 'llvm-ar' failed: No such file or directory
  ----------------------------------------
  ERROR: Failed building wheel for argon2-cffi
Failed to build argon2-cffi
ERROR: Could not build wheels for argon2-cffi, which is required to install pyproject.toml-based projects
~ $ pip3 install argon2-cffi
Collecting argon2-cffi
  Using cached argon2-cffi-21.1.0.tar.gz (1.8 MB)
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
Collecting cffi>=1.0.0
  Using cached cffi-1.15.0.tar.gz (484 kB)
  Preparing metadata (setup.py) ... done
Collecting pycparser
  Using cached pycparser-2.20-py2.py3-none-any.whl (112 kB)
Using legacy 'setup.py install' for cffi, since package 'wheel' is not installed.
Building wheels for collected packages: argon2-cffi
  Building wheel for argon2-cffi (pyproject.toml) ... error
  ERROR: Command errored out with exit status 1:
   command: /data/data/com.termux/files/usr/bin/python3 /data/data/com.termux/files/usr/lib/python3.10/site-packages/pip/_vendor/pep517/in_process/_in_process.py build_wheel /data/data/com.termux/files/usr/tmp/tmpc009lg3e
       cwd: /data/data/com.termux/files/usr/tmp/pip-install-6abwlbg3/argon2-cffi_2c454c38ef984db88dbcc253fbc27126
  Complete output (29 lines):
  running bdist_wheel
  running build
  running build_py
  creating build
  creating build/lib.linux-armv8l-3.10
  creating build/lib.linux-armv8l-3.10/argon2
  copying src/argon2/__init__.py -> build/lib.linux-armv8l-3.10/argon2
  copying src/argon2/low_level.py -> build/lib.linux-armv8l-3.10/argon2
  copying src/argon2/_ffi_build.py -> build/lib.linux-armv8l-3.10/argon2
  copying src/argon2/exceptions.py -> build/lib.linux-armv8l-3.10/argon2
  copying src/argon2/_utils.py -> build/lib.linux-armv8l-3.10/argon2
  copying src/argon2/_legacy.py -> build/lib.linux-armv8l-3.10/argon2
  copying src/argon2/_password_hasher.py -> build/lib.linux-armv8l-3.10/argon2
  copying src/argon2/__main__.py -> build/lib.linux-armv8l-3.10/argon2
  running build_clib
  building 'argon2' library
  creating build/temp.linux-armv8l-3.10
  creating build/temp.linux-armv8l-3.10/extras
  creating build/temp.linux-armv8l-3.10/extras/libargon2
  creating build/temp.linux-armv8l-3.10/extras/libargon2/src
  creating build/temp.linux-armv8l-3.10/extras/libargon2/src/blake2
  arm-linux-androideabi-clang -mfloat-abi=softfp -mfpu=vfpv3-d16 -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -march=armv7-a -mfpu=neon -mfloat-abi=softfp -mthumb -fstack-protector-strong -O3 -march=armv7-a -mfpu=neon -mfloat-abi=softfp -mthumb -fstack-protector-strong -O3 -fPIC -Iextras/libargon2/src/../include -Iextras/libargon2/src/blake2 -c extras/libargon2/src/argon2.c -o build/temp.linux-armv8l-3.10/extras/libargon2/src/argon2.o
  arm-linux-androideabi-clang -mfloat-abi=softfp -mfpu=vfpv3-d16 -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -march=armv7-a -mfpu=neon -mfloat-abi=softfp -mthumb -fstack-protector-strong -O3 -march=armv7-a -mfpu=neon -mfloat-abi=softfp -mthumb -fstack-protector-strong -O3 -fPIC -Iextras/libargon2/src/../include -Iextras/libargon2/src/blake2 -c extras/libargon2/src/blake2/blake2b.c -o build/temp.linux-armv8l-3.10/extras/libargon2/src/blake2/blake2b.o
  arm-linux-androideabi-clang -mfloat-abi=softfp -mfpu=vfpv3-d16 -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -march=armv7-a -mfpu=neon -mfloat-abi=softfp -mthumb -fstack-protector-strong -O3 -march=armv7-a -mfpu=neon -mfloat-abi=softfp -mthumb -fstack-protector-strong -O3 -fPIC -Iextras/libargon2/src/../include -Iextras/libargon2/src/blake2 -c extras/libargon2/src/core.c -o build/temp.linux-armv8l-3.10/extras/libargon2/src/core.o
  arm-linux-androideabi-clang -mfloat-abi=softfp -mfpu=vfpv3-d16 -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -march=armv7-a -mfpu=neon -mfloat-abi=softfp -mthumb -fstack-protector-strong -O3 -march=armv7-a -mfpu=neon -mfloat-abi=softfp -mthumb -fstack-protector-strong -O3 -fPIC -Iextras/libargon2/src/../include -Iextras/libargon2/src/blake2 -c extras/libargon2/src/encoding.c -o build/temp.linux-armv8l-3.10/extras/libargon2/src/encoding.o
  arm-linux-androideabi-clang -mfloat-abi=softfp -mfpu=vfpv3-d16 -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -march=armv7-a -mfpu=neon -mfloat-abi=softfp -mthumb -fstack-protector-strong -O3 -march=armv7-a -mfpu=neon -mfloat-abi=softfp -mthumb -fstack-protector-strong -O3 -fPIC -Iextras/libargon2/src/../include -Iextras/libargon2/src/blake2 -c extras/libargon2/src/ref.c -o build/temp.linux-armv8l-3.10/extras/libargon2/src/ref.o
  arm-linux-androideabi-clang -mfloat-abi=softfp -mfpu=vfpv3-d16 -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -march=armv7-a -mfpu=neon -mfloat-abi=softfp -mthumb -fstack-protector-strong -O3 -march=armv7-a -mfpu=neon -mfloat-abi=softfp -mthumb -fstack-protector-strong -O3 -fPIC -Iextras/libargon2/src/../include -Iextras/libargon2/src/blake2 -c extras/libargon2/src/thread.c -o build/temp.linux-armv8l-3.10/extras/libargon2/src/thread.o
  llvm-ar rcs build/temp.linux-armv8l-3.10/libargon2.a build/temp.linux-armv8l-3.10/extras/libargon2/src/argon2.o build/temp.linux-armv8l-3.10/extras/libargon2/src/blake2/blake2b.o build/temp.linux-armv8l-3.10/extras/libargon2/src/core.o build/temp.linux-armv8l-3.10/extras/libargon2/src/encoding.o build/temp.linux-armv8l-3.10/extras/libargon2/src/ref.o build/temp.linux-armv8l-3.10/extras/libargon2/src/thread.o
  error: command 'llvm-ar' failed: No such file or directory
  ----------------------------------------
  ERROR: Failed building wheel for argon2-cffi
Failed to build argon2-cffi
ERROR: Could not build wheels for argon2-cffi, which is required to install pyproject.toml-based projects

### What steps will reproduce the bug?

pip3 install jupyter

### System information

termux-info:
Application version:
0.117
Packages CPU architecture:
arm
Subscribed repositories:
# sources.list
deb https://termux.librehat.com/apt/termux-main/ stable main
# science-repo (sources.list.d/science.list)
deb https://packages.termux.org/apt/termux-science science stable
# game-repo (sources.list.d/game.list)
deb https://packages.termux.org/apt/termux-games games stable
Updatable packages:
All packages up to date
Android version:
9

**Comment by Grimler91** [comment]
So have you tried installing llvm to get llvm-ar?
```
$ $PREFIX/libexec/termux/command-not-found llvm-ar
The program llvm-ar is not installed. Install it by executing:
 pkg install llvm
```

**Comment by buttaface** [comment]
I wonder how it's calling `llvm-ar` in the first place: that Python package knows to call it on Android?

**Comment by Grimler91** [comment]
Probably comes from `$PREFIX/lib/python3.10/config-3.10/Makefile`:

```
$ grep -r llvm-ar $PREFIX/lib/python3.10
/data/data/com.termux/files/usr/lib/python3.10/_sysconfigdata__linux_.py: 'AR': 'llvm-ar',
grep: /data/data/com.termux/files/usr/lib/python3.10/__pycache__/_sysconfigdata__linux_.cpython-310.pyc: binary file matches
grep: /data/data/com.termux/files/usr/lib/python3.10/__pycache__/_sysconfigdata__linux_.cpython-310.opt-1.pyc: binary file matches
grep: /data/data/com.termux/files/usr/lib/python3.10/__pycache__/_sysconfigdata__linux_.cpython-310.opt-2.pyc: binary file matches
/data/data/com.termux/files/usr/lib/python3.10/config-3.10/Makefile:AR=		llvm-ar
```

**Comment by buttaface** [comment]
Yeah, we're going to have to do something about `ar` and`ranlib`, which were provided by the `binutils` dependency of `clang` before.

**Comment by buttaface** [comment]
I thought I'd have to patch the clang driver to use `llvm-ar`, [like I'll have to with Swift](https://github.com/apple/swift-driver/blob/9461e53ae5cb09aa938f66095a660b5630d4e4ef/Sources/SwiftDriver/Toolchains/GenericUnixToolchain.swift#L67), but clang just added a flag for static libraries last year, which seems to use llvm-ar already, llvm/llvm-project@4bafb0a.

Adding `llvm` as a dependency to clang will fix this bug, but not the cases where other build tools expect to use `ar`. Should we add a symlink, or is it so little used that we can just tell people to install `binutils` if they need it? I've added symlinks to `llvm-ar` and `llvm-ranlib` on my phone till I get the Swift compiler patched.

**Comment by buttaface** [comment]
Should be fixed by 7a89f6c, install the `llvm` package manually in the meantime.

**Comment by buttaface** [comment]
I looked into how CMake handles this and it turns out [it already uses `llvm-{ar,ranlib}` if the toolchain is clang](https://github.com/Kitware/CMake/blob/master/Modules/CMakeFindBinUtils.cmake#L175), like ours. I had to add those symbolic links when building LLVM with the Swift build-script on my phone because [the Swift build overrides those CMake tool variables](https://github.com/apple/swift/blob/main/utils/swift_build_support/swift_build_support/cmake.py#L157) for [the binutils' tool names](https://github.com/apple/swift/blob/main/utils/swift_build_support/swift_build_support/toolchain.py#L65), which it then can't find. I patched the tool names shown at that last link locally to use the LLVM versions instead, no problem and I no longer need the symlinks.

Since the LLVM tools will always be installed with clang after my latest revision bump, and clang, pip, and CMake all use the LLVM tools by default, closing this. I will patch Swift later and if anyone has problems with another compiler or build tool, please submit a patch.

**Comment by tomty89** [comment]
> Should we add a symlink, or is it so little used that we can just tell people to install binutils if they need it?

That's the problem -- both dependencies are barely appropriate and people can / should (make themselves capable enough to) just install them when necessary.

Whether it's common / of little use / capacity-consuming / relatively small does not really make it reasonable / unreasonable. It's not like you'll need it like you need a linker for even simple program. (Put it in another way, if what are talking about here is some metapackage that is supposed to pull in a reasonable toolchain, then the call made would have been appropriate -- even if it is made to pull both of them).

**Comment by tomty89** [comment]
Maybe I should put it this way (in case I'm wrong): does clang itself ever need the llvm subpackage to work? Or is it just what makes use of clang that does?

**Comment by buttaface** [comment]
> both dependencies are barely appropriate

You're saying we never should have had `binutils` as a clang dependency either?

> what we are talking about here is some metapackage that is supposed to pull in a reasonable toolchain, then the call made would have been appropriate

You are right that strictly speaking this is not needed to build a "simple program" and is more of a metapackage-like dependency. What I did with this NDK 23 update is replace the `binutils` dependency with the equivalent LLVM packages, like `lld` and `libcompiler-rt`, but I initially missed that `binutils` also provides the `ar` archiver, so now I added this `llvm` dependency for `llvm-ar`.

I simply replaced `binutils` with these three LLVM packages and was not thinking about whether it's a metapackage or not. If you think the way we've always done it is not the right approach, maybe submit a pull removing these clang dependencies and proposing some metapackage instead?

> does clang itself ever need the llvm subpackage to work? Or is it just what makes use of clang that does?

Both, I noted llvm/llvm-project@4bafb0a above, which added an `--emit-static-lib` flag to clang last year, which appears to invoke `llvm-ar` directly.
